File: src/test_runner/expect/toThrow.zig
pub fn toThrow(this: *Expect, globalThis: *JSGlobalObject, callFrame: *CallFrame) bun.JSError!JSValue {
    defer this.postMatch(globalThis);

    const thisValue = callFrame.this();
    const arguments = callFrame.argumentsAsArray(1);

    this.incrementExpectCallCounter();

    const expected_value: JSValue = brk: {
        const value = arguments[0];
        if (value.isUndefined()) {
            break :brk .zero;
        }
        if (value.isUndefinedOrNull() or !value.isObject() and !value.isString()) {
            var fmt = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
            return globalThis.throw("Expected value must be string or Error: {f}", .{value.toFmt(&fmt)});
        }
        if (value.isObject()) {
            if (ExpectAny.fromJSDirect(value)) |_| {
                if (ExpectAny.js.constructorValueGetCached(value)) |innerConstructorValue| {
                    break :brk innerConstructorValue;
                }
            }
        } else if (value.isString()) {
            // `.toThrow("") behaves the same as `.toThrow()`
            const s = try value.toJSString(globalThis);
            if (s.length() == 0) break :brk .zero;
        }
        break :brk value;
    };
    expected_value.ensureStillAlive();

    const not = this.flags.not;

    const result_, const return_value_from_function = try this.getValueAsToThrow(globalThis, try this.getValue(globalThis, thisValue, "toThrow", "<green>expected<r>"));

    const did_throw = result_ != null;

    if (not) {
        const signature = comptime getSignature("toThrow", "<green>expected<r>", true);

        if (!did_throw) return .js_undefined;

        const result: JSValue = result_.?;
        var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        defer formatter.deinit();

        if (expected_value == .zero or expected_value.isUndefined()) {
            const signature_no_args = comptime getSignature("toThrow", "", true);
            if (result.toError()) |err| {
                const name: JSValue = try err.getTruthyComptime(globalThis, "name") orelse .js_undefined;
                const message: JSValue = try err.getTruthyComptime(globalThis, "message") orelse .js_undefined;
                const fmt = signature_no_args ++ "\n\nError name: <red>{f}<r>\nError message: <red>{f}<r>\n";
                return globalThis.throwPretty(fmt, .{
                    name.toFmt(&formatter),
                    message.toFmt(&formatter),
                });
            }

            // non error thrown
            const fmt = signature_no_args ++ "\n\nThrown value: <red>{f}<r>\n";
            return globalThis.throwPretty(fmt, .{result.toFmt(&formatter)});
        }

        if (expected_value.isString()) {
            const received_message: JSValue = (if (result.isObject())
                try result.fastGet(globalThis, .message)
            else
                JSValue.fromCell(try result.toJSString(globalThis))) orelse .js_undefined;
            if (globalThis.hasException()) return .zero;

            // TODO: remove this allocation
            // partial match
            {
                const expected_slice = try expected_value.toSliceOrNull(globalThis);
                defer expected_slice.deinit();
                const received_slice = try received_message.toSliceOrNull(globalThis);
                defer received_slice.deinit();
                if (!strings.contains(received_slice.slice(), expected_slice.slice())) return .js_undefined;
            }

            return this.throw(globalThis, signature, "\n\nExpected substring: not <green>{f}<r>\nReceived message: <red>{f}<r>\n", .{
                expected_value.toFmt(&formatter),
                received_message.toFmt(&formatter),
            });
        }

        if (expected_value.isRegExp()) {
            const received_message: JSValue = (if (result.isObject())
                try result.fastGet(globalThis, .message)
            else
                JSValue.fromCell(try result.toJSString(globalThis))) orelse .js_undefined;

            if (globalThis.hasException()) return .zero;
            // TODO: REMOVE THIS GETTER! Expose a binding to call .test on the RegExp object directly.
            if (try expected_value.get(globalThis, "test")) |test_fn| {
                const matches = test_fn.call(globalThis, expected_value, &.{received_message}) catch |err| globalThis.takeException(err);
                if (!matches.toBoolean()) return .js_undefined;
            }

            return this.throw(globalThis, signature, "\n\nExpected pattern: not <green>{f}<r>\nReceived message: <red>{f}<r>\n", .{
                expected_value.toFmt(&formatter),
                received_message.toFmt(&formatter),
            });
        }

        if (try expected_value.fastGet(globalThis, .message)) |expected_message| {
            const received_message: JSValue = (if (result.isObject())
                try result.fastGet(globalThis, .message)
            else
                JSValue.fromCell(try result.toJSString(globalThis))) orelse .js_undefined;
            if (globalThis.hasException()) return .zero;

            // no partial match for this case
            if (!try expected_message.isSameValue(received_message, globalThis)) return .js_undefined;

            return this.throw(globalThis, signature, "\n\nExpected message: not <green>{f}<r>\n", .{expected_message.toFmt(&formatter)});
        }

        if (!result.isInstanceOf(globalThis, expected_value)) return .js_undefined;

        var expected_class = ZigString.Empty;
        try expected_value.getClassName(globalThis, &expected_class);
        const received_message: JSValue = (try result.fastGet(globalThis, .message)) orelse .js_undefined;
        return this.throw(globalThis, signature, "\n\nExpected constructor: not <green>{f}<r>\n\nReceived message: <red>{f}<r>\n", .{ expected_class, received_message.toFmt(&formatter) });
    }

    if (did_throw) {
        if (expected_value == .zero or expected_value.isUndefined()) return .js_undefined;

        const result: JSValue = if (result_.?.toError()) |r|
            r
        else
            result_.?;

        const _received_message: ?JSValue = if (result.isObject())
            try result.fastGet(globalThis, .message)
        else
            JSValue.fromCell(try result.toJSString(globalThis));

        if (expected_value.isString()) {
            if (_received_message) |received_message| {
                // TODO: remove this allocation
                // partial match
                const expected_slice = try expected_value.toSliceOrNull(globalThis);
                defer expected_slice.deinit();
                const received_slice = try received_message.toSlice(globalThis, globalThis.allocator());
                defer received_slice.deinit();
                if (strings.contains(received_slice.slice(), expected_slice.slice())) return .js_undefined;
            }

            // error: message from received error does not match expected string
            var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
            defer formatter.deinit();

            const signature = comptime getSignature("toThrow", "<green>expected<r>", false);

            if (_received_message) |received_message| {
                const expected_value_fmt = expected_value.toFmt(&formatter);
                const received_message_fmt = received_message.toFmt(&formatter);
                return this.throw(globalThis, signature, "\n\n" ++ "Expected substring: <green>{f}<r>\nReceived message: <red>{f}<r>\n", .{ expected_value_fmt, received_message_fmt });
            }

            const expected_fmt = expected_value.toFmt(&formatter);
            const received_fmt = result.toFmt(&formatter);
            return this.throw(globalThis, signature, "\n\n" ++ "Expected substring: <green>{f}<r>\nReceived value: <red>{f}<r>", .{ expected_fmt, received_fmt });
        }

        if (expected_value.isRegExp()) {
            if (_received_message) |received_message| {
                // TODO: REMOVE THIS GETTER! Expose a binding to call .test on the RegExp object directly.
                if (try expected_value.get(globalThis, "test")) |test_fn| {
                    const matches = test_fn.call(globalThis, expected_value, &.{received_message}) catch |err| globalThis.takeException(err);
                    if (matches.toBoolean()) return .js_undefined;
                }
            }

            // error: message from received error does not match expected pattern
            var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
            defer formatter.deinit();

            if (_received_message) |received_message| {
                const expected_value_fmt = expected_value.toFmt(&formatter);
                const received_message_fmt = received_message.toFmt(&formatter);
                const signature = comptime getSignature("toThrow", "<green>expected<r>", false);

                return this.throw(globalThis, signature, "\n\n" ++ "Expected pattern: <green>{f}<r>\nReceived message: <red>{f}<r>\n", .{ expected_value_fmt, received_message_fmt });
            }

            const expected_fmt = expected_value.toFmt(&formatter);
            const received_fmt = result.toFmt(&formatter);
            const signature = comptime getSignature("toThrow", "<green>expected<r>", false);
            return this.throw(globalThis, signature, "\n\n" ++ "Expected pattern: <green>{f}<r>\nReceived value: <red>{f}<r>", .{ expected_fmt, received_fmt });
        }

        if (Expect.isAsymmetricMatcher(expected_value)) {
            const signature = comptime getSignature("toThrow", "<green>expected<r>", false);
            const is_equal = try result.jestStrictDeepEquals(expected_value, globalThis);

            if (globalThis.hasException()) {
                return .zero;
            }

            if (is_equal) {
                return .js_undefined;
            }

            var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
            defer formatter.deinit();
            const received_fmt = result.toFmt(&formatter);
            const expected_fmt = expected_value.toFmt(&formatter);
            return this.throw(globalThis, signature, "\n\nExpected value: <green>{f}<r>\nReceived value: <red>{f}<r>\n", .{ expected_fmt, received_fmt });
        }

        // If it's not an object, we are going to crash here.
        assert(expected_value.isObject());

        if (try expected_value.fastGet(globalThis, .message)) |expected_message| {
            const signature = comptime getSignature("toThrow", "<green>expected<r>", false);

            if (_received_message) |received_message| {
                if (try received_message.isSameValue(expected_message, globalThis)) return .js_undefined;
            }

            // error: message from received error does not match expected error message.
            var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
            defer formatter.deinit();

            if (_received_message) |received_message| {
                const expected_fmt = expected_message.toFmt(&formatter);
                const received_fmt = received_message.toFmt(&formatter);
                return this.throw(globalThis, signature, "\n\nExpected message: <green>{f}<r>\nReceived message: <red>{f}<r>\n", .{ expected_fmt, received_fmt });
            }

            const expected_fmt = expected_message.toFmt(&formatter);
            const received_fmt = result.toFmt(&formatter);
            return this.throw(globalThis, signature, "\n\nExpected message: <green>{f}<r>\nReceived value: <red>{f}<r>\n", .{ expected_fmt, received_fmt });
        }

        if (result.isInstanceOf(globalThis, expected_value)) return .js_undefined;

        // error: received error not instance of received error constructor
        var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        defer formatter.deinit();
        var expected_class = ZigString.Empty;
        var received_class = ZigString.Empty;
        try expected_value.getClassName(globalThis, &expected_class);
        try result.getClassName(globalThis, &received_class);
        const signature = comptime getSignature("toThrow", "<green>expected<r>", false);
        const fmt = signature ++ "\n\nExpected constructor: <green>{f}<r>\nReceived constructor: <red>{f}<r>\n\n";

        if (_received_message) |received_message| {
            const message_fmt = fmt ++ "Received message: <red>{f}<r>\n";
            const received_message_fmt = received_message.toFmt(&formatter);

            return globalThis.throwPretty(message_fmt, .{
                expected_class,
                received_class,
                received_message_fmt,
            });
        }

        const received_fmt = result.toFmt(&formatter);
        const value_fmt = fmt ++ "Received value: <red>{f}<r>\n";

        return globalThis.throwPretty(value_fmt, .{
            expected_class,
            received_class,
            received_fmt,
        });
    }

    // did not throw
    const result = return_value_from_function;
    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();
    const received_line = "Received function did not throw\nReceived value: <red>{f}<r>\n";

    if (expected_value == .zero or expected_value.isUndefined()) {
        const signature = comptime getSignature("toThrow", "", false);
        return this.throw(globalThis, signature, "\n\n" ++ received_line, .{result.toFmt(&formatter)});
    }

    const signature = comptime getSignature("toThrow", "<green>expected<r>", false);

    if (expected_value.isString()) {
        const expected_fmt = "\n\nExpected substring: <green>{f}<r>\n\n" ++ received_line;
        return this.throw(globalThis, signature, expected_fmt, .{ expected_value.toFmt(&formatter), result.toFmt(&formatter) });
    }

    if (expected_value.isRegExp()) {
        const expected_fmt = "\n\nExpected pattern: <green>{f}<r>\n\n" ++ received_line;
        return this.throw(globalThis, signature, expected_fmt, .{ expected_value.toFmt(&formatter), result.toFmt(&formatter) });
    }

    if (try expected_value.fastGet(globalThis, .message)) |expected_message| {
        const expected_fmt = "\n\nExpected message: <green>{f}<r>\n\n" ++ received_line;
        return this.throw(globalThis, signature, expected_fmt, .{ expected_message.toFmt(&formatter), result.toFmt(&formatter) });
    }

    const expected_fmt = "\n\nExpected constructor: <green>{f}<r>\n\n" ++ received_line;
    var expected_class = ZigString.Empty;
    try expected_value.getClassName(globalThis, &expected_class);
    return this.throw(globalThis, signature, expected_fmt, .{ expected_class, result.toFmt(&formatter) });
}

const bun = @import("bun");
const ZigString = bun.ZigString;
const assert = bun.assert;
const strings = bun.strings;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;
const ExpectAny = bun.jsc.Expect.ExpectAny;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/test_runner/expect/toThrow.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
use bun_jsc::console_object::Formatter;
use bun_str::{strings, ZigString};

use super::Expect;
use super::expect_any::ExpectAny;
use super::get_signature;

#[bun_jsc::host_fn(method)]
pub fn to_throw(
    this: &mut Expect,
    global: &JSGlobalObject,
    frame: &CallFrame,
) -> JsResult<JSValue> {
    // TODO(port): `defer this.postMatch(globalThis)` — needs an RAII guard (e.g.
    // `let _post = this.post_match_guard(global);`) because a scopeguard capturing
    // `&mut self` here would conflict with later borrows. Phase B: add such a guard
    // on Expect or call post_match() before every return.

    let this_value = frame.this();
    let arguments = frame.arguments_as_array::<1>();

    this.increment_expect_call_counter();

    let expected_value: JSValue = 'brk: {
        let value = arguments[0];
        if value.is_undefined() {
            break 'brk JSValue::ZERO;
        }
        if value.is_undefined_or_null() || (!value.is_object() && !value.is_string()) {
            let mut fmt = Formatter::new(global).quote_strings(true);
            return global.throw(format_args!(
                "Expected value must be string or Error: {}",
                value.to_fmt(&mut fmt),
            ));
        }
        if value.is_object() {
            if ExpectAny::from_js_direct(value).is_some() {
                if let Some(inner_constructor_value) =
                    ExpectAny::js::constructor_value_get_cached(value)
                {
                    break 'brk inner_constructor_value;
                }
            }
        } else if value.is_string() {
            // `.toThrow("")` behaves the same as `.toThrow()`
            let s = value.to_js_string(global)?;
            if s.length() == 0 {
                break 'brk JSValue::ZERO;
            }
        }
        break 'brk value;
    };
    expected_value.ensure_still_alive();

    let not = this.flags.not;

    let (result_, return_value_from_function) = this.get_value_as_to_throw(
        global,
        this.get_value(global, this_value, "toThrow", "<green>expected<r>")?,
    )?;

    let did_throw = result_.is_some();

    if not {
        // PERF(port): was comptime — get_signature should be const fn returning &'static str
        let signature: &'static str = get_signature("toThrow", "<green>expected<r>", true);

        if !did_throw {
            return Ok(JSValue::UNDEFINED);
        }

        let result: JSValue = result_.unwrap();
        let mut formatter = Formatter::new(global).quote_strings(true);

        if expected_value.is_empty() || expected_value.is_undefined() {
            let signature_no_args: &'static str = get_signature("toThrow", "", true);
            if let Some(err) = result.to_error() {
                let name: JSValue = err
                    .get_truthy_comptime(global, "name")?
                    .unwrap_or(JSValue::UNDEFINED);
                let message: JSValue = err
                    .get_truthy_comptime(global, "message")?
                    .unwrap_or(JSValue::UNDEFINED);
                // TODO(port): comptime string concat — get_signature must be const fn for concatcp!
                let fmt = const_format::concatcp!(
                    get_signature("toThrow", "", true),
                    "\n\nError name: <red>{}<r>\nError message: <red>{}<r>\n"
                );
                return global.throw_pretty(
                    fmt,
                    format_args!(
                        "{}{}",
                        name.to_fmt(&mut formatter),
                        message.to_fmt(&mut formatter),
                    ),
                );
            }

            // non error thrown
            let fmt = const_format::concatcp!(
                get_signature("toThrow", "", true),
                "\n\nThrown value: <red>{}<r>\n"
            );
            return global.throw_pretty(fmt, format_args!("{}", result.to_fmt(&mut formatter)));
            // TODO(port): throw_pretty arg-threading — Zig passes a tuple matched to {f}
            // placeholders; Rust side likely wants format_args! directly. Revisit API shape.
            #[allow(unreachable_code)]
            let _ = signature_no_args;
        }

        if expected_value.is_string() {
            let received_message: JSValue = (if result.is_object() {
                result.fast_get(global, bun_jsc::BuiltinName::Message)?
            } else {
                Some(JSValue::from_cell(result.to_js_string(global)?))
            })
            .unwrap_or(JSValue::UNDEFINED);
            if global.has_exception() {
                return Ok(JSValue::ZERO);
            }

            // TODO: remove this allocation
            // partial match
            {
                let expected_slice = expected_value.to_slice_or_null(global)?;
                let received_slice = received_message.to_slice_or_null(global)?;
                if !strings::contains(received_slice.slice(), expected_slice.slice()) {
                    return Ok(JSValue::UNDEFINED);
                }
            }

            return this.throw(
                global,
                signature,
                "\n\nExpected substring: not <green>{}<r>\nReceived message: <red>{}<r>\n",
                format_args!(
                    "{}{}",
                    expected_value.to_fmt(&mut formatter),
                    received_message.to_fmt(&mut formatter),
                ),
            );
        }

        if expected_value.is_reg_exp() {
            let received_message: JSValue = (if result.is_object() {
                result.fast_get(global, bun_jsc::BuiltinName::Message)?
            } else {
                Some(JSValue::from_cell(result.to_js_string(global)?))
            })
            .unwrap_or(JSValue::UNDEFINED);

            if global.has_exception() {
                return Ok(JSValue::ZERO);
            }
            // TODO: REMOVE THIS GETTER! Expose a binding to call .test on the RegExp object directly.
            if let Some(test_fn) = expected_value.get(global, "test")? {
                let matches = test_fn
                    .call(global, expected_value, &[received_message])
                    .unwrap_or_else(|err| global.take_exception(err));
                if !matches.to_boolean() {
                    return Ok(JSValue::UNDEFINED);
                }
            }

            return this.throw(
                global,
                signature,
                "\n\nExpected pattern: not <green>{}<r>\nReceived message: <red>{}<r>\n",
                format_args!(
                    "{}{}",
                    expected_value.to_fmt(&mut formatter),
                    received_message.to_fmt(&mut formatter),
                ),
            );
        }

        if let Some(expected_message) = expected_value.fast_get(global, bun_jsc::BuiltinName::Message)? {
            let received_message: JSValue = (if result.is_object() {
                result.fast_get(global, bun_jsc::BuiltinName::Message)?
            } else {
                Some(JSValue::from_cell(result.to_js_string(global)?))
            })
            .unwrap_or(JSValue::UNDEFINED);
            if global.has_exception() {
                return Ok(JSValue::ZERO);
            }

            // no partial match for this case
            if !expected_message.is_same_value(received_message, global)? {
                return Ok(JSValue::UNDEFINED);
            }

            return this.throw(
                global,
                signature,
                "\n\nExpected message: not <green>{}<r>\n",
                format_args!("{}", expected_message.to_fmt(&mut formatter)),
            );
        }

        if !result.is_instance_of(global, expected_value) {
            return Ok(JSValue::UNDEFINED);
        }

        let mut expected_class = ZigString::EMPTY;
        expected_value.get_class_name(global, &mut expected_class)?;
        let received_message: JSValue = result
            .fast_get(global, bun_jsc::BuiltinName::Message)?
            .unwrap_or(JSValue::UNDEFINED);
        return this.throw(
            global,
            signature,
            "\n\nExpected constructor: not <green>{}<r>\n\nReceived message: <red>{}<r>\n",
            format_args!("{}{}", expected_class, received_message.to_fmt(&mut formatter)),
        );
    }

    if did_throw {
        if expected_value.is_empty() || expected_value.is_undefined() {
            return Ok(JSValue::UNDEFINED);
        }

        let result: JSValue = if let Some(r) = result_.unwrap().to_error() {
            r
        } else {
            result_.unwrap()
        };

        let received_message_opt: Option<JSValue> = if result.is_object() {
            result.fast_get(global, bun_jsc::BuiltinName::Message)?
        } else {
            Some(JSValue::from_cell(result.to_js_string(global)?))
        };

        if expected_value.is_string() {
            if let Some(received_message) = received_message_opt {
                // TODO: remove this allocation
                // partial match
                let expected_slice = expected_value.to_slice_or_null(global)?;
                let received_slice = received_message.to_slice(global)?;
                if strings::contains(received_slice.slice(), expected_slice.slice()) {
                    return Ok(JSValue::UNDEFINED);
                }
            }

            // error: message from received error does not match expected string
            let mut formatter = Formatter::new(global).quote_strings(true);

            let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);

            if let Some(received_message) = received_message_opt {
                let expected_value_fmt = expected_value.to_fmt(&mut formatter);
                let received_message_fmt = received_message.to_fmt(&mut formatter);
                return this.throw(
                    global,
                    signature,
                    concat!("\n\n", "Expected substring: <green>{}<r>\nReceived message: <red>{}<r>\n"),
                    format_args!("{}{}", expected_value_fmt, received_message_fmt),
                );
            }

            let expected_fmt = expected_value.to_fmt(&mut formatter);
            let received_fmt = result.to_fmt(&mut formatter);
            return this.throw(
                global,
                signature,
                concat!("\n\n", "Expected substring: <green>{}<r>\nReceived value: <red>{}<r>"),
                format_args!("{}{}", expected_fmt, received_fmt),
            );
        }

        if expected_value.is_reg_exp() {
            if let Some(received_message) = received_message_opt {
                // TODO: REMOVE THIS GETTER! Expose a binding to call .test on the RegExp object directly.
                if let Some(test_fn) = expected_value.get(global, "test")? {
                    let matches = test_fn
                        .call(global, expected_value, &[received_message])
                        .unwrap_or_else(|err| global.take_exception(err));
                    if matches.to_boolean() {
                        return Ok(JSValue::UNDEFINED);
                    }
                }
            }

            // error: message from received error does not match expected pattern
            let mut formatter = Formatter::new(global).quote_strings(true);

            if let Some(received_message) = received_message_opt {
                let expected_value_fmt = expected_value.to_fmt(&mut formatter);
                let received_message_fmt = received_message.to_fmt(&mut formatter);
                let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);

                return this.throw(
                    global,
                    signature,
                    concat!("\n\n", "Expected pattern: <green>{}<r>\nReceived message: <red>{}<r>\n"),
                    format_args!("{}{}", expected_value_fmt, received_message_fmt),
                );
            }

            let expected_fmt = expected_value.to_fmt(&mut formatter);
            let received_fmt = result.to_fmt(&mut formatter);
            let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);
            return this.throw(
                global,
                signature,
                concat!("\n\n", "Expected pattern: <green>{}<r>\nReceived value: <red>{}<r>"),
                format_args!("{}{}", expected_fmt, received_fmt),
            );
        }

        if Expect::is_asymmetric_matcher(expected_value) {
            let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);
            let is_equal = result.jest_strict_deep_equals(expected_value, global)?;

            if global.has_exception() {
                return Ok(JSValue::ZERO);
            }

            if is_equal {
                return Ok(JSValue::UNDEFINED);
            }

            let mut formatter = Formatter::new(global).quote_strings(true);
            let received_fmt = result.to_fmt(&mut formatter);
            let expected_fmt = expected_value.to_fmt(&mut formatter);
            return this.throw(
                global,
                signature,
                "\n\nExpected value: <green>{}<r>\nReceived value: <red>{}<r>\n",
                format_args!("{}{}", expected_fmt, received_fmt),
            );
        }

        // If it's not an object, we are going to crash here.
        debug_assert!(expected_value.is_object());

        if let Some(expected_message) = expected_value.fast_get(global, bun_jsc::BuiltinName::Message)? {
            let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);

            if let Some(received_message) = received_message_opt {
                if received_message.is_same_value(expected_message, global)? {
                    return Ok(JSValue::UNDEFINED);
                }
            }

            // error: message from received error does not match expected error message.
            let mut formatter = Formatter::new(global).quote_strings(true);

            if let Some(received_message) = received_message_opt {
                let expected_fmt = expected_message.to_fmt(&mut formatter);
                let received_fmt = received_message.to_fmt(&mut formatter);
                return this.throw(
                    global,
                    signature,
                    "\n\nExpected message: <green>{}<r>\nReceived message: <red>{}<r>\n",
                    format_args!("{}{}", expected_fmt, received_fmt),
                );
            }

            let expected_fmt = expected_message.to_fmt(&mut formatter);
            let received_fmt = result.to_fmt(&mut formatter);
            return this.throw(
                global,
                signature,
                "\n\nExpected message: <green>{}<r>\nReceived value: <red>{}<r>\n",
                format_args!("{}{}", expected_fmt, received_fmt),
            );
        }

        if result.is_instance_of(global, expected_value) {
            return Ok(JSValue::UNDEFINED);
        }

        // error: received error not instance of received error constructor
        let mut formatter = Formatter::new(global).quote_strings(true);
        let mut expected_class = ZigString::EMPTY;
        let mut received_class = ZigString::EMPTY;
        expected_value.get_class_name(global, &mut expected_class)?;
        result.get_class_name(global, &mut received_class)?;
        let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);
        // TODO(port): comptime string concat — requires get_signature to be const fn
        let fmt = const_format::concatcp!(
            get_signature("toThrow", "<green>expected<r>", false),
            "\n\nExpected constructor: <green>{}<r>\nReceived constructor: <red>{}<r>\n\n"
        );

        if let Some(received_message) = received_message_opt {
            let message_fmt = const_format::concatcp!(fmt, "Received message: <red>{}<r>\n");
            let received_message_fmt = received_message.to_fmt(&mut formatter);

            return global.throw_pretty(
                message_fmt,
                format_args!("{}{}{}", expected_class, received_class, received_message_fmt),
            );
        }

        let received_fmt = result.to_fmt(&mut formatter);
        let value_fmt = const_format::concatcp!(fmt, "Received value: <red>{}<r>\n");

        return global.throw_pretty(
            value_fmt,
            format_args!("{}{}{}", expected_class, received_class, received_fmt),
        );
        #[allow(unreachable_code)]
        let _ = signature;
    }

    // did not throw
    let result = return_value_from_function;
    let mut formatter = Formatter::new(global).quote_strings(true);
    // PORT NOTE: Zig `received_line` was concatenated via `++` into each fmt string
    // below; Rust `concat!` only accepts literals so the value is inlined at each site.
    // received_line = "Received function did not throw\nReceived value: <red>{any}<r>\n"

    if expected_value.is_empty() || expected_value.is_undefined() {
        let signature: &'static str = get_signature("toThrow", "", false);
        return this.throw(
            global,
            signature,
            concat!("\n\n", "Received function did not throw\nReceived value: <red>{}<r>\n"),
            format_args!("{}", result.to_fmt(&mut formatter)),
        );
    }

    let signature: &'static str = get_signature("toThrow", "<green>expected<r>", false);

    if expected_value.is_string() {
        let expected_fmt = concat!(
            "\n\nExpected substring: <green>{}<r>\n\n",
            "Received function did not throw\nReceived value: <red>{}<r>\n"
        );
        return this.throw(
            global,
            signature,
            expected_fmt,
            format_args!(
                "{}{}",
                expected_value.to_fmt(&mut formatter),
                result.to_fmt(&mut formatter),
            ),
        );
    }

    if expected_value.is_reg_exp() {
        let expected_fmt = concat!(
            "\n\nExpected pattern: <green>{}<r>\n\n",
            "Received function did not throw\nReceived value: <red>{}<r>\n"
        );
        return this.throw(
            global,
            signature,
            expected_fmt,
            format_args!(
                "{}{}",
                expected_value.to_fmt(&mut formatter),
                result.to_fmt(&mut formatter),
            ),
        );
    }

    if let Some(expected_message) = expected_value.fast_get(global, bun_jsc::BuiltinName::Message)? {
        let expected_fmt = concat!(
            "\n\nExpected message: <green>{}<r>\n\n",
            "Received function did not throw\nReceived value: <red>{}<r>\n"
        );
        return this.throw(
            global,
            signature,
            expected_fmt,
            format_args!(
                "{}{}",
                expected_message.to_fmt(&mut formatter),
                result.to_fmt(&mut formatter),
            ),
        );
    }

    let expected_fmt = concat!(
        "\n\nExpected constructor: <green>{}<r>\n\n",
        "Received function did not throw\nReceived value: <red>{}<r>\n"
    );
    let mut expected_class = ZigString::EMPTY;
    expected_value.get_class_name(global, &mut expected_class)?;
    return this.throw(
        global,
        signature,
        expected_fmt,
        format_args!("{}{}", expected_class, result.to_fmt(&mut formatter)),
    );
}

// TODO(port): confirm `bun_jsc::BuiltinName::Message` is the correct variant
// path for `fastGet(global, .message)` once `bun_jsc` exposes the builtin-name enum.

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/test_runner/expect/toThrow.zig (319 lines)
//   confidence: medium
//   todos:      5
//   notes:      throw_pretty/this.throw fmt-arg threading is approximated via format_args!; get_signature must be const fn for concatcp!; defer post_match needs RAII guard; BuiltinName::Message path unconfirmed
// ──────────────────────────────────────────────────────────────────────────
